File: cpp/sim7000.cpp
#include "inc/sim7000.hpp"

#include <wiringPi.h>
#include <wiringSerial.h>

#include <algorithm>
#include <iostream>
#include <unistd.h>
#include <string.h>
#include <chrono>
#include <thread>

SIM7000::SIM7000(std::string _port, const int _baudrate, const bool _postInitOn, const int _powerKey) : port(_port),
                                                                                                        baudrate(_baudrate),
                                                                                                        postInitOn(_postInitOn),
                                                                                                        powerKey(_powerKey)
{
    this->fd = 0;
    this->lastMessage = "";
    this->lastCommand = "";
    this->lastInfo = "";

    if (this->postInitOn)
    {
        this->connectToSerial();
        if (this->isConnected())
        {
            this->setGPIO();
            this->simOn();
        }
    }
}

SIM7000::~SIM7000()
{
    this->simOff();
    this->disconnectSerial();
    this->cleanGPIO();
}

void SIM7000::connectToSerial()
{
    this->fd = serialOpen(this->port.c_str(), this->baudrate);
}

void SIM7000::disconnectSerial()
{
    serialClose(this->fd);
    this->fd = 0;
}

const bool SIM7000::isConnected()
{
    if (this->fd > 0)
    {
        return true;
    }

    return false;
}

void SIM7000::setGPIO()
{
    wiringPiSetupGpio();
    pinMode(this->powerKey, OUTPUT);
}

void SIM7000::cleanGPIO()
{
    digitalWrite(this->powerKey, LOW);
}

void SIM7000::simOn()
{
    digitalWrite(this->powerKey, HIGH);
    std::this_thread::sleep_for(std::chrono::milliseconds(1000));
    digitalWrite(this->powerKey, LOW);
    std::this_thread::sleep_for(std::chrono::milliseconds(1000));

    int num = 0;
    while (num < 10 && this->lastInfo != "OK")
    {
        num++;
        this->sendMessage();
        std::cout << "Trying to turn on SIM7000 for " << num << " time/s. Last info: " << this->lastInfo << "!\n";
    }
}

void SIM7000::simOff()
{
    digitalWrite(this->powerKey, HIGH);
    std::this_thread::sleep_for(std::chrono::milliseconds(1000));
    digitalWrite(this->powerKey, LOW);
    std::this_thread::sleep_for(std::chrono::milliseconds(1000));
}

std::string SIM7000::getLastMessage()
{
    return this->lastMessage;
}

std::string SIM7000::getLastCommand()
{
    return this->lastCommand;
}

std::string SIM7000::getLastInfo()
{
    return this->lastInfo;
}

std::string SIM7000::sendMessage(std::string command, const int bufferSize)
{
    command += '\r';
    serialFlush(this->fd);
    std::this_thread::sleep_for(std::chrono::milliseconds(1200));
    serialPuts(this->fd, command.c_str());
    std::this_thread::sleep_for(std::chrono::milliseconds(1200));

    char buffer[bufferSize];
    memset(buffer, '\0', bufferSize);
    read(this->fd, buffer, bufferSize);

    std::string bufferString = buffer;
    this->lastMessage = bufferString;

    int numberOfr = 0;
    std::string temporaryCommand = "";
    std::string temporaryInfo = "";
    for (char s : this->lastMessage)
    {
        switch (s)
        {
        case '\r':
            numberOfr++;
            break;
        case '\n':
            break;
        default:
            if (numberOfr < 2)
            {
                temporaryCommand += s;
            }
            else
            {
                temporaryInfo += s;
            }
            break;
        }
    }
    this->lastCommand = temporaryCommand;
    this->lastInfo = temporaryInfo;

    return this->lastMessage;
}

std::vector<std::string> SIM7000::sendListMessages(std::vector<std::string> commands, const int bufferSize)
{
    std::vector<std::string> listMessages;

    for (std::string cmd : commands)
    {
        cmd += '\r';
        serialFlush(this->fd);
        std::this_thread::sleep_for(std::chrono::milliseconds(1200));
        serialPuts(this->fd, cmd.c_str());
        std::this_thread::sleep_for(std::chrono::milliseconds(1200));

        char buffer[bufferSize];
        memset(buffer, '\0', bufferSize);
        read(this->fd, buffer, bufferSize);

        std::string bufferString = buffer;
        this->lastMessage = bufferString;
        listMessages.emplace_back(this->lastMessage);
    }
    int numberOfr = 0;
    std::string temporaryCommand = "";
    std::string temporaryInfo = "";
    for (char s : this->lastMessage)
    {
        switch (s)
        {
        case '\r':
            numberOfr++;
            break;
        case '\n':
            break;
        default:
            if (numberOfr < 2)
            {
                temporaryCommand += s;
            }
            else
            {
                temporaryInfo += s;
            }
            break;
        }
    }
    this->lastCommand = temporaryCommand;
    this->lastInfo = temporaryInfo;

    return listMessages;
}

File: python/sim7000.py
import serial
import time
import RPi.GPIO as GPIO
from dataclasses import dataclass
from typing import List, NoReturn, Union, Optional


@dataclass
class SIM7000:
    port: str = '/dev/serial0'
    baudrate: int = 115200
    bytesize: int = serial.EIGHTBITS
    timeout: int = 1
    stopbits: Union[int, float] = serial.STOPBITS_ONE
    parity: str = serial.PARITY_NONE
    xonxoff: bool = True
    power_key: int = 4

    post_init_on: bool = True

    __ser: Optional[serial.Serial] = None
    __last_message: Optional[str] = None
    __last_command: Optional[str] = None
    __last_info: Optional[str] = None

    def __post_init__(self) -> NoReturn:
        if(self.post_init_on):
            self._connect_to_serial()
            if(self._is_connected()):
                self.__set_gpio()
                self.__sim_on()

    def __del__(self) -> NoReturn:
        self.__sim_off()
        self._disconnet_serial()
        self.__clean_gpio()

    def _connect_to_serial(self) -> NoReturn:
        self.__ser = serial.Serial(port=self.port,
                                   baudrate=self.baudrate,
                                   bytesize=self.bytesize,
                                   timeout=self.timeout,
                                   stopbits=self.stopbits,
                                   parity=self.parity,
                                   xonxoff=self.xonxoff)

    def _disconnet_serial(self) -> NoReturn:
        self.__ser.close()

    def _is_connected(self) -> bool:
        return self.__ser.is_open

    def __set_gpio(self) -> NoReturn:
        GPIO.setmode(GPIO.BCM)
        GPIO.setwarnings(False)
        GPIO.setup(self.power_key, GPIO.OUT)

    def __clean_gpio(self) -> NoReturn:
        GPIO.cleanup()

    def __sim_on(self) -> NoReturn:
        GPIO.output(self.power_key, GPIO.HIGH)
        time.sleep(1)
        GPIO.output(self.power_key, GPIO.LOW)
        time.sleep(1)

        num = 0
        while(num < 10 and self.__last_info != 'OK'):
            num += 1
            self.send_message()
            print(
                f'\nTrying to turn on SIM7000 for {num} time/s. Last info: {self.__last_info}!')

    def __sim_off(self) -> NoReturn:
        GPIO.output(self.power_key, GPIO.HIGH)
        time.sleep(1)
        GPIO.output(self.power_key, GPIO.LOW)
        time.sleep(1)

    @property
    def get_last_message(self) -> Optional[str]:
        return self.__last_message

    @property
    def get_last_command(self) -> Optional[str]:
        return self.__last_command

    @property
    def get_last_info(self) -> Optional[str]:
        return self.__last_info

    def send_message(self, command: str = 'at', buff_size: int = 128) -> str:
        self.__ser.write(bytes(command+'\r', 'utf-8'))
        self.__last_message = self.__ser.read(buff_size).decode('utf-8')
        self.__last_command = self.__last_message.replace(
            '\r\n', '').split('\r')[0]
        self.__last_info = self.__last_message.replace(
            '\r\n', '').split('\r')[1]

        return self.__last_message

    def send_list_messages(self, list_commands: List[str] = ['at'], buff_size: int = 128) -> List[str]:
        list_messages = []

        for cmd in list_commands:
            self.__ser.write(bytes(cmd+'\r', 'utf-8'))
            self.__last_message = self.__ser.read(buff_size).decode('utf-8')
            list_messages.append(self.__last_message)
        self.__last_command = self.__last_message.replace(
            '\r\n', '').split('\r')[0]
        self.__last_info = self.__last_message.replace(
            '\r\n', '').split('\r')[1]

        return list_messages
